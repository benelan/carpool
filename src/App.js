import React, { Component } from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import "./index.css";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { loadModules } from "esri-loader";
import { Spinner } from "reactstrap";
import Header from "./components/Header";
import Settings from "./components/Settings";
import ResultTable from "./components/ResultTable";

class App extends Component {

  state = {
    name: 1,
    email: null
  }

    //--------------------- LIFE CYCLE FUNCTIONS ---------------------\\
    componentDidMount() {
      loadModules(["esri/portal/Portal", "esri/identity/OAuthInfo", "esri/identity/IdentityManager"])
        .then(([Portal, OAuthInfo, IdentityManager]) => {
          var info = new OAuthInfo({
            appId: "n5A1575tmQq5eFPd",
            popup: false
          });
  
          var that = this;
          
          IdentityManager.registerOAuthInfos([info]);
          IdentityManager.getCredential(info.portalUrl + "/sharing");
          IdentityManager.checkSignInStatus(info.portalUrl + "/sharing")
            .then((e) => {
              var portal = new Portal();
              // Setting authMode to immediate signs the user in once loaded
              portal.authMode = "immediate";
              // Once loaded, user is signed in
              portal.load().then(function () {
                // set state and form of email and name
                that.setState({
                  name: portal.user.fullName,
                  email: portal.user.email
                });
              });  
            })
           
        })
        .catch(err => {
          // handle any errors
          console.error(err);
        });
        console.log(this.state)
    };
  
    componentWillUnmount() {
      loadModules(["esri/identity/OAuthInfo", "esri/identity/IdentityManager"])
        .then(([OAuthInfo, IdentityManager]) => {
          var info = new OAuthInfo({
            appId: "n5A1575tmQq5eFPd",
            popup: false
          });
          IdentityManager.registerOAuthInfos([info]);
          IdentityManager.destroyCredentials();
        });
       
      window.location.reload();
    }
  

  render() {
    const center = {
      // margin: "auto",
      // padding: "1px",
      // width: "50%",
      // left: "50%",
      bottom: "50px"
    };

    const { name, email } = this.state
    return !!email  && !!name ? 
    <div>
    <Header />
    <Router>

          <Switch>
            <Route exact path="/settings">
              <Settings n={this.state.name} e={this.state.email} />
            </Route>
            <Route exact path="/results">
              <ResultTable n={this.state.name} e={this.state.email} />
            </Route>
          </Switch>

      </Router>
      </div>
    : (
      <div className="center" >
      <Spinner color="warning" style={{ width: '3rem', height: '3rem' }} />
      </div>
    )
  }
}
export default App;
