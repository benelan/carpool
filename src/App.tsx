import React from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import "./index.css";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import { loadModules } from "esri-loader";
import { Spinner } from "reactstrap";
import Header from "./components/Header";
import Settings from "./components/Settings";
import Home from "./components/Home";
import ResultTable from "./components/ResultTable";

type MyState = {
  email: string | null,
  name: string | null;
};

class App extends React.Component<MyState> {

  state: MyState = {
    name: null,
    email: null
  }

  //--------------------- LIFE CYCLE FUNCTIONS ---------------------\\
  async componentDidMount(): Promise<void> {
    // load modules
    type MapModules = [
      typeof import("esri/portal/Portal"),
      typeof import("esri/identity/OAuthInfo"),
      typeof import("esri/identity/IdentityManager")
    ];

    const [Portal, OAuthInfo, IdentityManager] = await (loadModules([
      "esri/portal/Portal",
      "esri/identity/OAuthInfo",
      "esri/identity/IdentityManager"
    ]) as Promise<MapModules>);


    var info = new OAuthInfo({
      appId: "n5A1575tmQq5eFPd",
      popup: false
    });

    var that = this;

    IdentityManager.registerOAuthInfos([info]);
    IdentityManager.getCredential(info.portalUrl + "/sharing");
    IdentityManager.checkSignInStatus(info.portalUrl + "/sharing")
      .then(() => {
        var portal = new Portal();
        // Setting authMode to immediate signs the user in once loaded
        portal.authMode = "immediate";
        // Once loaded, user is signed in
        portal.load().then(function () {
          // set state and form of email and name
          that.setState({
            name: portal.user.fullName,
            email: portal.user.email
          });
        });
      })
  };

  async componentWillUnmount(): Promise<void> {
    // load modules
    type MapModules = [
      typeof import("esri/identity/OAuthInfo"),
      typeof import("esri/identity/IdentityManager")
    ];
    const [OAuthInfo, IdentityManager] = await (loadModules([
      "esri/identity/OAuthInfo",
      "esri/identity/IdentityManager"
    ]) as Promise<MapModules>);

    // destroy credentials
    var info = new OAuthInfo({
      appId: "n5A1575tmQq5eFPd",
      popup: false
    });
    IdentityManager.registerOAuthInfos([info]);
    IdentityManager.destroyCredentials();

    // reload
    window.location.reload();
  }


  render() {
    return (
      <Router>
        <Header />
        <div>{!!this.state.email && !!this.state.name ?

          <Switch>
            <Route exact path="/">
              <Home n={this.state.name} e={this.state.email} />
            </Route>
            <Route exact path="/settings">
              <Settings n={this.state.name} e={this.state.email} />
            </Route>
            <Route exact path="/results">
              <ResultTable n={this.state.name} e={this.state.email} />
            </Route>
          </Switch>

          : (
            <div className="center" >
              <Spinner color="warning" style={{ width: '3rem', height: '3rem' }} />
            </div>
          )}</div>
      </Router>
    )
  }
}
export default App;
