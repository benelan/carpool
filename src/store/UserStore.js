import { action, observable, decorate } from "mobx"

class UserStore {
    userName = null;
    userEmail = null;
    officeId = 1;
    driver = 1;
    arrive = "09:00";
    leave = "17:00";
    successful = false;
    userNew = true;
    lineId = null;
    pointId = null;
    route = null;
    address = '';
    loaded = false;
    offsite = false;

    setName(x) {
        this.userName = x;
    }
    setEmail(x) {
        this.userEmail = x;
    }
    setNew(x) {
        this.userNew = x;
    }
    setDriver(x) {
        this.driver = x;
    }
    setArrive(x) {
        this.arrive = x;
    }
    setLeave(x) {
        this.leave = x;
    }
    setOffice(x) {
        this.officeId = x;
    }
    setSuccess(x) {
        this.successful = x;
    }
    setLineId(x) {
        this.lineId = x;
    }
    setPointId(x) {
        this.pointId = x;
    }
    setRoute(x) {
        this.route = x;
    }
    setAddress(x) {
        this.address = x;
    }
    setLoaded(x) {
        this.loaded = x;
    }
    setOffsite(x) {
        this.offsite = x;
    }
}

  
decorate(UserStore, {
    userNew: observable,
    userName: observable,
    userEmail: observable,
    officeId: observable,
    leave: observable,
    arrive: observable,
    driver: observable,
    successful: observable,
    lineId: observable,
    pointId: observable,
    route: observable,
    address: observable,
    loaded: observable,
    offsite: observable,
    setNew: action,
    setName: action,
    setEmail: action,
    setDriver: action,
    setOffice: action,
    setLeave: action,
    setArrive: action,
    setSuccess: action,
    setLineId: action,
    setPointId: action,
    setRoute: action,
    setAddress: action,
    setLoaded: action,
    setOffsite: action
})

export default UserStore