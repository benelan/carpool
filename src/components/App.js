import React from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import "../index.css";
import { BrowserRouter as Router, Switch, Route, useLocation, Redirect } from "react-router-dom";
import { Provider } from 'mobx-react'
import UserStore from '../store/userDomainStore'
import Header from "./Header";
import Settings from "./Settings";
import Home from "./Home";
import Login from "./Login";
import ResultTable from "./ResultTable";

// A custom hook that builds on useLocation to parse
// the query string for you.
function useQuery() {
  return new URLSearchParams(useLocation().search);
}

const App = () => {
  return (
    <Router>
      <Provider UserStore={UserStore}>
        <Routes />
      </Provider>
    </Router>
  );
}

const Routes = () => {
  let query = useQuery();
  return (
    <React.Fragment>
      <Header />
      <Switch>
        <Redirect exact from="/" to="/home" />
        <Route exact path="/home">
          <Home />
        </Route>
        <Route exact path="/settings">
          <Settings />
        </Route>
        <Route exact path="/results">
          <ResultTable />
        </Route>
        <Route exact path="/login">
          <Login n={query.get("name")} e={query.get("email")} />
        </Route>
      </Switch>
    </React.Fragment>
  )
}

export default App