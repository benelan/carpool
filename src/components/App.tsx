import React from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import "../index.css";
import { BrowserRouter as Router, Switch, Route, Redirect } from "react-router-dom";
import { loadModules } from "esri-loader";
import { Spinner } from "reactstrap";
import Header from "./Header";
import Settings from "./Settings";
import Home from "./Home";
import ResultTable from "./ResultTable";
import { observer, inject } from 'mobx-react'
import UserStore from '../store/UserStore';
import axios from "axios";


type MyProps = {
  UserStore?: UserStore
}

const App = inject("UserStore")(observer(
  class App extends React.Component<MyProps> {

    CancelToken = axios.CancelToken;
    source = this.CancelToken.source();
    proxyUrl: string = 'https://belan2.esri.com/DotNet/proxy.ashx?';


    //--------------------- LIFE CYCLE FUNCTIONS ---------------------\\
    async componentDidMount(): Promise<void> {
      // load modules
      type MapModules = [
        typeof import("esri/portal/Portal"),
        typeof import("esri/identity/OAuthInfo"),
        typeof import("esri/identity/IdentityManager")
      ];

      const [Portal, OAuthInfo, IdentityManager] = await (loadModules([
        "esri/portal/Portal",
        "esri/identity/OAuthInfo",
        "esri/identity/IdentityManager"
      ]) as Promise<MapModules>);


      var info = new OAuthInfo({
        appId: "n5A1575tmQq5eFPd",
        popup: false
      });

      var that = this;

      IdentityManager.registerOAuthInfos([info]);
      IdentityManager.getCredential(info.portalUrl + "/sharing");
      IdentityManager.checkSignInStatus(info.portalUrl + "/sharing")
        .then(() => {
          var portal = new Portal();
          // Setting authMode to immediate signs the user in once loaded
          portal.authMode = "immediate";
          // Once loaded, user is signed in
          portal.load().then(function () {
            // set state and form of email and name
            that.props.UserStore!.setName(portal.user.fullName)
            that.props.UserStore!.setEmail(portal.user.email)

            that.getUserByEmail();
          });
        })
    };

    async componentWillUnmount(): Promise<void> {
      // load modules
      type MapModules = [
        typeof import("esri/identity/OAuthInfo"),
        typeof import("esri/identity/IdentityManager")
      ];
      const [OAuthInfo, IdentityManager] = await (loadModules([
        "esri/identity/OAuthInfo",
        "esri/identity/IdentityManager"
      ]) as Promise<MapModules>);

      // destroy credentials
      var info = new OAuthInfo({
        appId: "n5A1575tmQq5eFPd",
        popup: false
      });
      IdentityManager.registerOAuthInfos([info]);
      IdentityManager.destroyCredentials();

      // reload
      window.location.reload();
    }

    async getUserByEmail(): Promise<void> {
      //------------------------------------------ POINT ------------------------------------------\\
      const serviceUrl: string = 'https://services.arcgis.com/Wl7Y1m92PbjtJs5n/arcgis/rest/services/carpoolData/FeatureServer/0/query?';

      let url: string = this.proxyUrl + serviceUrl;

      const data: any = {
        "f": "json",
        'where': "email='" + this.props.UserStore!.userEmail + "'",
        'outFields': "*",
        'timestamp': new Date().getTime()
      };

      const query: string = Object.keys(data)
        .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(data[k]))
        .join('&');

      url = url + query;

      await axios.get(url)
        .then(res => {
          const users: Array<any> = res.data.features;
          if (users.length > 0) {  // check to see if user is already saved
            const user: any = users[0].attributes;
            // populate form with user data
            this.props.UserStore!.setArrive(user.arrive_work)
            this.props.UserStore!.setLeave(user.leave_work)
            this.props.UserStore!.setDriver(user.driver)
            this.props.UserStore!.setOffice(user.office_id)
            this.props.UserStore!.setSuccess(!!user.success)
            this.props.UserStore!.setNew(false);
            this.props.UserStore!.setPointId(user.OBJECTID);
            this.props.UserStore!.setAddress(user.start_addr);
          }
          this.props.UserStore!.setLoaded(true);
        })
        .catch(err => {
          this.props.UserStore!.setLoaded(true);
          this.props.UserStore!.setOffsite(true);
          console.log(err)
        });
      //------------------------------------------ Line ------------------------------------------\\
      const serviceUrl2: string = 'https://services.arcgis.com/Wl7Y1m92PbjtJs5n/arcgis/rest/services/carpoolData/FeatureServer/1/query?';
      let url2: string = this.proxyUrl + serviceUrl2;

      const data2: any = {
        "f": "json",
        'where': "email='" + this.props.UserStore!.userEmail + "'",
        'outFields': "*",
        'returnGeometry': true,
        'timestamp': new Date().getTime()
      };

      const query2: string = Object.keys(data2)
        .map(k => encodeURIComponent(k) + '=' + encodeURIComponent(data2[k]))
        .join('&');

      url2 = url2 + query2;

      await axios.get(url2)
        .then(res => {
          const users: Array<any> = res.data.features;
          // fill in form and state with settings saved in db
          if (users.length > 0) {  // check to see if user is already saved
            const user: any = users[0];
            // populate form with user data
            this.props.UserStore!.setLineId(user.attributes.OBJECTID)
            this.props.UserStore!.setRoute({
              spatialReference: res.data.spatialReference,
              paths: user.geometry.paths,
              type: 'polyline'
            })
          }
        })
        .catch(err => {
          console.log(err)
        });
    };

    render() {
      return (
        <Router>
          <Header />
          {!!this.props.UserStore!.userEmail && this.props.UserStore!.loaded ?
            <Switch>
              <Redirect exact from="/" to="/home" />
              <Route exact path="/home">
                <Home />
              </Route>
              <Route exact path="/settings">
                <Settings />
              </Route>
              <Route exact path="/results">
                <ResultTable />
              </Route>
            </Switch>
            : (
              <div className="center" >
                <Spinner color="warning" style={{ width: '3rem', height: '3rem' }} />
              </div>
            )}
        </Router>
      )
    }
  }))
export default App;
