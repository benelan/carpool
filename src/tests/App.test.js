  
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from '../components/App';
import UserStore from "../store/UserStore"
import { Provider } from 'mobx-react';

// basic render
it('App renders without crashing', () => {
  const store = new UserStore()
  const div = document.createElement('div');
  ReactDOM.render(<Provider UserStore={store}><App /></Provider>, div);
});


// snapshot
it("App snapshot", () => {
  const store = new UserStore()
  const wrapper = shallow(
    <Provider UserStore={store}><App /></Provider>
  );
  expect(wrapper).toMatchSnapshot();
});