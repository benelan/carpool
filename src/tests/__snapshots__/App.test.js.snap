// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<BrowserRouter>
  <MobXProvider
    UserStore={
      UserStore {
        "arrive": "09:00",
        "driver": 1,
        "leave": "17:00",
        "officeId": 1,
        "successful": false,
        "userEmail": null,
        "userName": null,
        "userNew": true,
      }
    }
  >
    <Routes />
  </MobXProvider>
</BrowserRouter>
`;
